### DISCUSSION TITLE
How to obtain the course angle from the simulation

### DISCUSSION THREAD

**Discussion by Dobid** [JSBSim-Team/jsbsim#852] [Q&A]
Hello, me again ! I'm still trying to implement a PID controller for the [x8](https://github.com/AOS55/Fixedwing-Airsim/blob/master/x8/x8.xml) UAV.

In the [book](https://drive.google.com/file/d/19xEFoIhGGz5ftc8r89Z6pHlUZHI3bPE4/view?usp=sharing) I'm following, they introduce the notion of successive loop closure. As showed in the book, I'm trying to implement an inner loop controlling the roll angle by commanding the ailerons, and an outer loop controlling the course angle by commanding the roll angle.

My problem is that I can't find in JSBSim, a property that returns the "course angle" denoted $\chi$ as described in these figures from the book : 
![image](https://user-images.githubusercontent.com/36502022/223179495-f904aa4d-d2c0-4d4c-a3a1-d601df386da5.png)

I've tried `attitude/psi-rad`, `attitude/heading-true-rad` and `flight-path/psi-gt-rad`. When plotted, all of them seem to be the same and representing the yaw angle staying at zero (which I can't control, I only have elevons).
![image](https://user-images.githubusercontent.com/36502022/223182181-97ea184a-2dbb-4dda-a3f8-f3b793abd8b3.png)

Is there a way of retrieving this course angle $\chi$ ?

Thanks a lot !

**Comment by seanmcleod** [comment]
Isn't $\chi$ simply based on the angle with regards to the North and East velocities?

```c++
  PropertyManager->Tie("velocities/v-north-fps", this, eNorth, (PMF)&FGPropagate::GetVel);
  PropertyManager->Tie("velocities/v-east-fps", this, eEast, (PMF)&FGPropagate::GetVel);
```

**Reply by Dobid** [reply]
You're right, it is.

Now, what are the differences between the 3 properties I mentionned in my question ? They all seem to be the same.

**Comment by seanmcleod** [comment]
> Now, what are the differences between the 3 properties I mentionned in my question ?

Think about the wind.

```c++
PropertyManager->Tie("flight-path/psi-gt-rad", this, &FGAuxiliary::GetGroundTrack);

double GetGroundTrack(void)        const { return psigt;         }

psigt = atan2(in.vVel(eEast), in.vVel(eNorth));
if (psigt < 0.0) psigt += 2*M_PI;
```

```c++
PropertyManager->Tie("attitude/psi-rad", this, (int)ePsi, (PMF)&FGPropagate::GetEuler);

  /** Retrieves a vehicle Euler angle component.
      Retrieves an Euler angle (Phi, Theta, or Psi) from the quaternion that
      stores the vehicle orientation relative to the Local frame. The order of
      rotations used is Yaw-Pitch-Roll. The Euler angle with subscript (1) is
      Phi. Various convenience enumerators are defined in FGJSBBase. The
      relevant enumerators for the Euler angle returned by this call are,
      ePhi=1, eTht=2, ePsi=3 (e.g. GetEuler(eTht) returns Theta).
      units radians
      @return An Euler angle.
  */
  double GetEuler(int axis) const { return VState.qAttitudeLocal.GetEuler(axis); }
```

```c++
PropertyManager->Tie("attitude/heading-true-rad", this, (int)ePsi, (PMF)&FGPropagate::GetEuler);
```

**Reply by Dobid** [reply]
Not sure about what you mean about the wind. However thanks to the pointers to the code you showed I'm able to understand better what those properties return :

https://github.com/JSBSim-Team/jsbsim/blob/b2172a493c0bcbba2c27b6149c4e3cd3b53f8983/src/FGFDMExec.cpp#L481

https://github.com/JSBSim-Team/jsbsim/blob/b2172a493c0bcbba2c27b6149c4e3cd3b53f8983/src/models/FGPropagate.h#L285-L296

So `psi-gt-rad` indeed corresponds to $\chi$ right ? If the wind blows, the 'drift' induced is taken into account since the frame of reference is the inertial NED frame.

And since they both call to the same `GetEuler` method, `psi-rad` and `heading-true-rad` are both the same euler angle ?

**Comment by seanmcleod** [comment]
> So psi-gt-rad indeed corresponds to $\chi$ right ? 

Yep.

And as you can see in the code snippets `attitude/psi-rad` and `attitude/heading-true-rad` are always identical.

Now if there is no wind then `flight-path/psi-gt-rad` will also be the same. But if there is some cross-wind component then `flight-path/psi-gt-rad` will differ compared to the other two.

**Reply by Dobid** [reply]
Alright I got it, very clear now. Thanks !
